cpp/src/arrow/json/parser.cc

```diff
@@ -0,0 +1,982 @@
+      case Kind::kArray:
+        return Cast<Kind::kArray>(builder_)->AppendNull();
+      case Kind::kObject: {
+        auto root = builder_;
+        auto struct_builder = Cast<Kind::kObject>(builder_);
+        RETURN_NOT_OK(struct_builder->AppendNull());
+        for (int i = 0; i != struct_builder->num_fields(); ++i) {
+          builder_ = struct_builder->field_builder(i);
+          RETURN_NOT_OK(AppendNull());
+        }
+        builder_ = root;
+        return Status::OK();
+      }
+      default:
+        return Status::NotImplemented("invalid builder Kind");
+    }
+  }
+
+  Status AppendBool(bool value) {
+    constexpr auto kind = Kind::kBoolean;
+    if (builder_.kind != kind) {
+      return IllegallyChangedTo(kind);
+    }
+    return Cast<kind>(builder_)->Append(value);
+  }
+
+  template <Kind::type kind>
+  Status AppendScalar(string_view scalar) {
+    if (builder_.kind != kind) {
+      return IllegallyChangedTo(kind);
+    }
+    auto index = static_cast<int32_t>(scalar_values_builder_.length());
+    RETURN_NOT_OK(Cast<kind>(builder_)->Append(index));
+    return scalar_values_builder_.Append(scalar);
+  }
+
+  Status StartObjectImpl() {
+    constexpr auto kind = Kind::kObject;
+    if (builder_.kind != kind) {
+      return IllegallyChangedTo(kind);
+    }
+    auto struct_builder = Cast<kind>(builder_);
+    absent_fields_stack_.Push(struct_builder->num_fields(), true);
+    PushStacks();
+    return struct_builder->Append();
+  }
+
+  bool SetFieldBuilder(string_view key) {
+    auto parent = Cast<Kind::kObject>(builder_stack_.back());
+    field_index_ = parent->GetFieldIndex(std::string(key));
+    if (field_index_ == -1) {
+      return false;
+    }
+    builder_ = parent->field_builder(field_index_);
+    absent_fields_stack_[field_index_] = false;
+    return true;
+  }
+
+  Status EndObjectImpl() {
+    auto parent = Cast<Kind::kObject>(builder_stack_.back());
+
+    auto expected_count = absent_fields_stack_.TopSize();
+    for (field_index_ = 0; field_index_ != expected_count; ++field_index_) {
+      if (!absent_fields_stack_[field_index_]) {
+        continue;
+      }
+      builder_ = parent->field_builder(field_index_);
+      if (!builder_.nullable) {
+        return ParseError("a required field was absent");
+      }
+      RETURN_NOT_OK(AppendNull());
+    }
+    absent_fields_stack_.Pop();
+    PopStacks();
+    return Status::OK();
+  }
+
+  Status StartArrayImpl() {
+    constexpr auto kind = Kind::kArray;
+    if (builder_.kind != kind) {
+      return IllegallyChangedTo(kind);
+    }
+    PushStacks();
+    // append to the list builder in EndArrayImpl
+    builder_ = Cast<kind>(builder_)->value_builder();
+    return Status::OK();
+  }
+
+  Status EndArrayImpl(rapidjson::SizeType size) {
+    PopStacks();
+    // append to list_builder here
+    auto list_builder = Cast<Kind::kArray>(builder_);
+    return list_builder->Append(size);
+  }
+
+  void PushStacks() {
```

Comment:
What does this function do, and the next one?

Reply:
I have added some comments

Reply:
PushStacks() adds the current builder to the stack so its children can be visited and parsed. PopStacks() replaces the current builder with its parent when parsing of its children has been completed, so that parsing of the parent can continue. I could rename them `{Push,Pop}CurrentBuilder()` if that's more clear